lib/samlp.js

```diff
@@ -180,6 +238,105 @@ function generateInstant() {
   return date.getUTCFullYear() + '-' + ('0' + (date.getUTCMonth()+1)).slice(-2) + '-' + ('0' + date.getUTCDate()).slice(-2) + 'T' + ('0' + date.getUTCHours()).slice(-2) + ":" + ('0' + date.getUTCMinutes()).slice(-2) + ":" + ('0' + date.getUTCSeconds()).slice(-2) + "Z";
 }
 
+// Validates an XML that contains a Signature element (signature inside XML)
+function validateXmlSignature(xml, options, callback) {
+  var calculatedThumbprint = '';
+  var signature = options.signature || xpath.select(signaturePath, xml)[0];
+  if (!signature){
+    return callback(new Error('Signature is missing (xpath: ' + signaturePath + ')'));
+  }
+
+  if (options.thumprints){
+    // Make sure thumprints is an array for the validation 
+    options.thumbprints = options.thumbprints instanceof Array ? options.thumbprints : [options.thumbprints];
+  }
+
+  var sig = new SignedXml(null, { idAttribute: 'AssertionID' });
+
+  sig.keyInfoProvider = {
+    getKeyInfo: function () {
+      return "<X509Data></X509Data>";
+    },
+    getKey: function (keyInfo) {
+
+      //If there's no embedded signing cert, use the configured cert through options
+      if(!keyInfo || keyInfo.length===0){
+        if(!options.cert) throw new Error('options.cert must be specified for SAMLResponses with no embedded signing certificate');
+        return certToPEM(options.cert);
+      }
+
+      //If there's an embedded signature and thumprints are provided check that
+      if (options.thumbprints && options.thumbprints.length > 0)  {
+        var embeddedSignature = keyInfo[0].getElementsByTagNameNS("http://www.w3.org/2000/09/xmldsig#", "X509Certificate");
+        if (embeddedSignature.length > 0) {
+          var base64cer = embeddedSignature[0].firstChild.toString();
+          var shasum = crypto.createHash('sha1');
+          var der = new Buffer(base64cer, 'base64').toString('binary');
+          shasum.update(der);
+          calculatedThumbprint = shasum.digest('hex');
+          
+          // using embedded cert, so options.cert is not used anymore
+          delete options.cert;
```

Comment:
Because we will use the embedded certificate to confirm the signature and then the thumbprint to validate the certificate